framework/src/kernels/ADKernel.C

```diff
@@ -138,36 +138,63 @@ ADKernelTempl<T>::computeResidual()
 
 template <typename T>
 void
-ADKernelTempl<T>::computeJacobian()
+ADKernelTempl<T>::computeResidualsForJacobian()
 {
-  prepareMatrixTag(_assembly, _var.number(), _var.number());
-
-  auto ad_offset =
-      Moose::adOffset(_var.number(), _sys.getMaxVarNDofsPerElem(), Moose::ElementType::Element);
+  if (_residuals.size() != _test.size())
+    _residuals.resize(_test.size(), 0);
+  for (auto & r : _residuals)
+    r = 0;
 
   precalculateResidual();
-
   if (_use_displaced_mesh)
-    for (_i = 0; _i < _test.size(); _i++)
-      for (_qp = 0; _qp < _qrule->n_points(); _qp++)
-      {
-        DualReal residual = _ad_JxW[_qp] * _ad_coord[_qp] * computeQpResidual();
-        for (_j = 0; _j < _var.phiSize(); _j++)
-          _local_ke(_i, _j) += residual.derivatives()[ad_offset + _j];
-      }
+    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    {
+      _r = _ad_JxW[_qp];
+      _r *= _ad_coord[_qp];
+      for (_i = 0; _i < _test.size(); _i++)
+        _residuals[_i] += _r * computeQpResidual();
+    }
   else
-    for (_i = 0; _i < _test.size(); _i++)
-      for (_qp = 0; _qp < _qrule->n_points(); _qp++)
-      {
-        DualReal residual = _JxW[_qp] * _coord[_qp] * computeQpResidual();
-        for (_j = 0; _j < _var.phiSize(); _j++)
-          _local_ke(_i, _j) += residual.derivatives()[ad_offset + _j];
-      }
+    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+      for (_i = 0; _i < _test.size(); _i++)
+        _residuals[_i] += _JxW[_qp] * _coord[_qp] * computeQpResidual();
+}
+
+template <typename T>
+void
+ADKernelTempl<T>::addJacobian(const MooseVariableFieldBase & jvariable)
+{
+  unsigned int jvar = jvariable.number();
+
+  auto ad_offset = Moose::adOffset(jvar, _sys.getMaxVarNDofsPerElem(), Moose::ElementType::Element);
+
+  prepareMatrixTag(_assembly, _var.number(), jvar);
+
+  for (_i = 0; _i < _test.size(); _i++)
+    for (_j = 0; _j < jvariable.phiSize(); _j++)
+      _local_ke(_i, _j) += _residuals[_i].derivatives()[ad_offset + _j];
 
   accumulateTaggedLocalMatrix();
+}
+
+template <typename T>
+void
+ADKernelTempl<T>::computeJacobian()
+{
+  computeResidualsForJacobian();
+
+  auto local_functor = [&](const std::vector<ADReal> &,
+                           const std::vector<dof_id_type> &,
+                           const std::set<TagID> &) { addJacobian(_var); };
+
+  _assembly.processDerivatives(_residuals, _var.dofIndices(), _matrix_tags, local_functor);
 
   if (_has_diag_save_in && !_sys.computingScalingJacobian())
   {
+#ifdef MOOSE_GLOBAL_AD_INDEXING
```

Comment:
same here - it would be nice to keep this logic out of kernels, bcs, etc.  But maybe it's not worth the effort for such a low-level class like ADKernel.